library/std/src/sys/windows/process.rs

```diff
@@ -31,32 +32,72 @@ use libc::{c_void, EXIT_FAILURE, EXIT_SUCCESS};
 // Command
 ////////////////////////////////////////////////////////////////////////////////
 
-#[derive(Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Clone, Debug, Eq)]
 #[doc(hidden)]
-pub struct EnvKey(OsString);
+pub struct EnvKey {
+    os_string: OsString,
+    utf16: Vec<u16>,
+}
+
+// Windows environment variables preserve their case but comparisions use
+// simplified case folding. So we call `CompareStringOrdinal` to get the OS to
+// perform the comparision.
+impl Ord for EnvKey {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        unsafe {
+            let result = c::CompareStringOrdinal(
+                self.utf16.as_ptr(),
+                self.utf16.len() as _,
+                other.utf16.as_ptr(),
+                other.utf16.len() as _,
+                c::TRUE,
+            );
+            match result {
+                c::CSTR_LESS_THAN => cmp::Ordering::Less,
+                c::CSTR_EQUAL => cmp::Ordering::Equal,
+                c::CSTR_GREATER_THAN => cmp::Ordering::Greater,
+                // `CompareStringOrdinal` should never fail so long as the parameters are correct.
+                _ => panic!("comparing environment keys failed: {}", Error::last_os_error()),
+            }
+        }
+    }
+}
+impl PartialOrd for EnvKey {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl PartialEq for EnvKey {
+    fn eq(&self, other: &Self) -> bool {
+        if self.utf16.len() != other.utf16.len() {
+            false
+        } else {
+            self.cmp(other) == cmp::Ordering::Equal
+        }
+    }
+}
 
 impl From<OsString> for EnvKey {
-    fn from(mut k: OsString) -> Self {
-        k.make_ascii_uppercase();
-        EnvKey(k)
+    fn from(k: OsString) -> Self {
+        EnvKey { utf16: k.encode_wide().collect(), os_string: k }
```

Comment:
There is a helper function`sys::windows::to_u16s`. It might be possible to use that instead and remove the `ensure_no_nuls` in this module.
https://github.com/rust-lang/rust/blob/3bcaeb0bf9e1c29d18abc32928fd2f23d1bed0bd/library/std/src/sys/windows/mod.rs#L139-L152

Reply:
My thinking was that the `From` implementation does not allow returning an error whereas `make_envp` does. So I'd have to turn it into a panic in place of the error?

Reply:
Mmm yeah that's true. Then I would leave it as is, refactoring things to reduce duplication can always be done in a later as it is not really related to this PR.